### ISSUE TITLE
Bugfix for calling AvroConvert.OpenDeserializer with only one record in a datablock

### ISSUE THREAD

**Issue by steinbohrer** [AdrianStrugala/AvroConvert#59]
### Problem:
I'm not allowed to share the Avro file which I used, but just create/pick a file with only one record in the first datablock (type: record) and try to iterate over it using AvroConvert.OpenDeserializer.
You will most likely get an Exception.

For me, the Exception occured in the ResolveUnion() method of the Resolver class:

```
            int index = d.ReadUnionIndex();
            TypeSchema ws = writerSchema.Schemas[index];

```
The call to ReadUnionIndex() returned a 7 but the Schemas array only contained 2 items.
So, what went wrong?

### Explanation:
When you try to re-create the scenario and run it you will end up in "ILineReader<T> OpenReader<T>()" in the Decoder class on the line:

`return new ListLineReader<T>(reader, new Resolver(writeSchema, readSchema));`

because "remainingBlocks == 1" and the type is record and not array. Btw, "remainingBlocks" is actually the number of records in the current datablock, so you should consider renaming it.

Be aware that the reader object passed to ListLineReader is a new MemoryStream that ONLY contains the actual data.

In the constructor of ListLineReader a call is always made to ReadArrayStart():

`itemsCount = (int)reader.ReadArrayStart();`

In my case this has read a "1" from the datablock, which was the correct value to pick from the Schemas array.

### Solution:
To fix this, I changed the constructor of ListLineReader to make the call to ReadArrayStart() optional:

```
        private long itemsCount;

        public ListLineReader(IReader reader, Resolver resolver, long itemsCount = -1)
        {
            this.reader = reader;
            this.resolver = resolver;

            this.itemsCount = itemsCount;
            if (itemsCount == -1)
            {
                this.itemsCount = reader.ReadArrayStart();
            }
        }
```

You can remove the cast to int because ReadArrayStart() returns a long.

And in the OpenReader method of the Decoder class add a 1 to both calls to ListLineReader because we know that "remainingBlocks" is 1.

```
                if (writeSchema.Type == AvroType.Array) {
                    return new ListLineReader<T>(reader, new Resolver(((ArraySchema)writeSchema).ItemSchema, readSchema), 1);
                }

                return new ListLineReader<T>(reader, new Resolver(writeSchema, readSchema), 1);
```

The first call to ListLineReader, when the Avro file does not contain a header does not pass a value to itemsCount and so the call to ReadArrayStart() is still done. But I have not tested that scenario.

Doing this will fix the problem that the first long value of the datablock will always be taken as an itemsCount, even though it should not.

**Comment by steinbohrer** [comment]
Oh, btw. to prevent the reader from reading past the end of the current datablock, you can add the fololowing to "IReader.cs":

```
        /// <summary>
        /// Test if the underlying stream still has data to read, or not.
        /// </summary>
        /// <returns>False if all data has been read from the underlying stream.</returns>
        bool HasData();
```

and implement it in "Reader.cs":

```
        public bool HasData()
        {
            return stream.Position < streamLength;
        }
```

Then, use this in the ListLineReader HasNext() method:

```
        public bool HasNext()
        {
            if (itemsCount == 0)
            {
                if (reader.HasData())
                {
                    itemsCount = reader.ReadArrayNext();
                }
                return itemsCount != 0;
            }
            else
            {
                return true;
            }
        }

```

**Comment by AdrianStrugala** [comment]
Hello,

Thanks for pointing out the problems and proposing a fix. I'm reworking completely the 'DeserializeByLine' feature (according also to your previous posts) and it should be ready in few days. If you spot any more problems, please let me know.

Thanks,
Adrian

**Comment by AdrianStrugala** [comment]
Hello,

The bug is resolved. The fix is ready in 3.2.1 version of AvroConvert

Thanks,
Adrian
